File: boat-common/src/test/java/test/java/xyz/srclab/common/proxy/ProxyTest.java
package test.java.xyz.srclab.common.proxy;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.testng.Assert;
import org.testng.annotations.Test;
import xyz.srclab.common.base.Current;
import xyz.srclab.common.proxy.*;
import xyz.srclab.common.test.TestLogger;

import java.lang.reflect.Method;
import java.util.Arrays;

public class ProxyTest {

    private static final TestLogger logger = TestLogger.DEFAULT;

    @Test
    public void testProxy() {
        doTestProxy(JdkProxyClassFactory.INSTANCE);
    }

    private void doTestProxy(ProxyClassFactory proxyClassFactory) {
        ProxyMethod<TP> proxyMethod = new ProxyMethod<TP>() {
            @NotNull
            @Override
            public String name() {
                return "hello";
            }

            @Override
            public Class<?>[] parametersTypes() {
                return new Class[]{String.class, String.class};
            }

            @Override
            public Object invoke(
                    TP proxy, @NotNull Method method, @Nullable Object[] args, @NotNull SuperInvoker superInvoker) {
                logger.log("method: {}, declaring class: {}", method, method.getDeclaringClass());
                return "proxy-> " + superInvoker.invoke(args);
            }
        };
        ProxyClass<TP> proxyClass = ProxyClass.newProxyClass(
                TP.class, Arrays.asList(proxyMethod), Current.classLoader(), proxyClassFactory);
        Assert.assertEquals(proxyClass.newInstance().hello("a", "b"), "proxy-> hello: a = a, b = b");
    }

    public static class TP {

        public String hello(String a, String b) {
            return "hello: a = " + a + ", b = " + b;
        }
    }
}

File: easy-common-proxy/src/test/kotlin/test/xyz/scrlab/common/proxy/ProxyTest.kt
package test.xyz.scrlab.common.proxy

import org.testng.annotations.Test
import xyz.srclab.common.proxy.ClassProxy
import xyz.srclab.common.proxy.provider.ClassProxyProvider
import xyz.srclab.common.proxy.provider.bytecode.ByteCodeClassProxyProvider
import xyz.srclab.common.proxy.provider.jdk.JdkClassProxyProvider
import xyz.srclab.common.reflect.invoke.MethodInvoker
import xyz.srclab.common.reflect.method.ProxyMethod
import xyz.srclab.test.doAssertEquals
import java.lang.reflect.Method

/**
 * @author sunqian
 */
object ProxyTest {

    @Test
    fun testProxy() {
        val classProxy = ClassProxy.newBuilder(A::class.java)
            .proxyMethod("hello", arrayOf(), object : ProxyMethod {
                override fun invoke(`object`: Any, args: Array<Any>, method: Method, superInvoker: MethodInvoker): Any {
                    return "proxy hello"
                }
            })
            .build()
        val a = classProxy.newInstance()
        doAssertEquals(a.hello(), "proxy hello")
    }

    @Test
    fun testWithProvider() {
        doTestProxy(JdkClassProxyProvider.INSTANCE, A::class.java)
        doTestProxy(ByteCodeClassProxyProvider.INSTANCE, A::class.java)
    }

    private fun <T : A> doTestProxy(provider: ClassProxyProvider, baseClass: Class<T>) {
        val classProxy = provider.newBuilder(baseClass)
            .proxyMethod("hello", arrayOf(), object : ProxyMethod {
                override fun invoke(`object`: Any, args: Array<Any>, method: Method, superInvoker: MethodInvoker): Any {
                    return "proxy hello"
                }
            })
            .build()
        val a = classProxy.newInstance()
        doAssertEquals(a.hello(), "proxy hello")
    }

    interface A {
        fun hello(): String
    }
}